prediction/src/algorithms/segment/src/training.py

```diff
@@ -0,0 +1,88 @@
+import glob
+import math
+import os
+
+import numpy as np
+import pylidc as pl
+from keras.callbacks import ModelCheckpoint
+from keras.models import load_model
+from src.preprocess.lung_segmentation import save_lung_segments
+
+from .data_generation import get_dicom_paths
+from .model import simple_model_3d
+
+
+def get_data_shape():
+    dicom_paths = get_dicom_paths()
+    max_x, max_y, max_z = get_max_scaled_dimensions(dicom_paths)
+    return 128, 128, 128, 1  # TODO: dynamically crop shape based on maximal axes
+    return max_x, max_y, max_z, 1
+
+
+def get_best_model_path():
+    current_dir = os.path.dirname(os.path.realpath(__file__))
+    assets_dir = os.path.abspath(os.path.join(current_dir, '../assets'))
+    return os.path.join(assets_dir, 'best_model.hdf5')
+
+
+def get_max_scaled_dimensions(dicom_paths):
+    """Return the maximum x, y and z dimensions of all scaled DICOM images"""
+    max_x, max_y, max_z = 0, 0, 0
+    for path in dicom_paths:
+        print(path)
+        directories = path.split('/')
+        lidc_id = directories[2]
+        scan = pl.query(pl.Scan).filter(pl.Scan.patient_id == lidc_id).first()
+        dicoms = scan.load_all_dicom_images()
+        scaled_x, scaled_y = dicoms[0].Rows * scan.pixel_spacing, dicoms[0].Columns * scan.pixel_spacing
+        scaled_z = len(dicoms) * scan.slice_thickness
+        scaled_x, scaled_y, scaled_z = math.ceil(scaled_x), math.ceil(scaled_y), math.ceil(scaled_z)
+
+        if scaled_x > max_x:
+            max_x = scaled_x
+        if scaled_y > max_y:
+            max_y = scaled_y
+        if scaled_z > max_z:
+            max_z = scaled_z
+    return max_x, max_y, max_z
+
+
+def train(load_checkpoint=False):
+    """Load the training masks from the asset folder and train a keras model"""
+    CUBOID_IMAGE_SHAPE = get_data_shape()
+    CUBOID_BATCH = 1  # How many training pairs should be passed to model.fit in one batch
+
+    current_dir = os.path.dirname(os.path.realpath(__file__))
+    assets_dir = os.path.abspath(os.path.join(current_dir, '../assets'))
+    dicom_paths = get_dicom_paths()
+
+    labels = glob.glob(os.path.join(assets_dir, "segmented_lung_patient_*.npy"))
+
+    input_data = np.zeros((CUBOID_BATCH, *CUBOID_IMAGE_SHAPE))
+    output_data = np.zeros((CUBOID_BATCH, *CUBOID_IMAGE_SHAPE))
+    best_model_path = get_best_model_path()
+
+    if os.path.isfile(best_model_path) and load_checkpoint:
+        model = load_model(best_model_path)
+    else:
+        model = simple_model_3d(CUBOID_IMAGE_SHAPE)
+
+    for batch_index in range(0, len(labels), CUBOID_BATCH):
+
+        for index, path in enumerate(dicom_paths[batch_index:batch_index + CUBOID_BATCH]):
+            lidc_id = path.split('/')[5]
+            patient_id = path.split('/')[-1]
```

Comment:
minor nitpick, but how about setting the result of `path.split('/')` to a variable so that the `split` method is only called once?

Reply:
Also, take a look at `os.path.splitext`, `os.path.split`, or `PurePath.parts` which should be more platform independent than `split('/')`. If none of those work, use [`os.path.sep`](https://stackoverflow.com/a/16595356/408556) instead of hardcoding `/`.

Reply:
Thanks, implemented in 28660bf681dad4acae8e20ee15433d8a48884384